JoinStressTest.testTCPIPJoinWithManyNodes

com.hazelcast.cluster.JoinStressTest.testTCPIPJoinWithManyNodes

```
java.lang.AssertionError: expected:<20> but was:<18>
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.failNotEquals(Assert.java:743)
    at org.junit.Assert.assertEquals(Assert.java:118)
    at org.junit.Assert.assertEquals(Assert.java:555)
    at org.junit.Assert.assertEquals(Assert.java:542)
    at com.hazelcast.cluster.JoinStressTest.testJoinWithManyNodes(JoinStressTest.java:108)
    at com.hazelcast.cluster.JoinStressTest.testTCPIPJoinWithManyNodes(JoinStressTest.java:68)
```

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/345/testReport/junit/com.hazelcast.cluster/JoinStressTest/testTCPIPJoinWithManyNodes/

--- a/hazelcast/src/main/java/com/hazelcast/cluster/Joiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/cluster/Joiner.java
@@ -35,16 +35,34 @@ public interface Joiner {
     /**
      * Adds an address to the blacklist. Blacklist is useful if a node should ignore another node, e.g. when
      * the groupname of 2 machines is not the same and they should form different clusters.
-     *
+     * <p/>
+     * If blacklist is permanent, then this operation is write-once. It cannot be unblacklisted again.
+     * If blacklist is temporary, blacklist can be removed via {@link #unblacklist(com.hazelcast.nio.Address)}.
+     * <p/>
      * Method is thread-safe.
-     *
+     * <p/>
      * If the address already is blacklisted, the call is ignored
      *
-     * @param address the address to blacklist.
+     * @param address   the address to blacklist.
+     * @param permanent if blacklist is permanent or not
      * @throws java.lang.NullPointerException if address is null.
      * @see #isBlacklisted(com.hazelcast.nio.Address)
      */
-    void blacklist(Address address);
+    void blacklist(Address address, boolean permanent);
+
+    /**
+     * Removes an address from the blacklist if it's temporarily blacklisted.
+     * This method has no effect if given address is not blacklisted. Permanent blacklists
+     * cannot be undone.
+     * <p/>
+     * Method is thread-safe.
+     * <p/>
+     * If the address is not blacklisted, the call is ignored.
+     *
+     * @param address the address to unblacklist.
+     * @return true if address is unblacklisted, false otherwise.
+     */
+    boolean unblacklist(Address address);
 
     /**
      * Checks if an address is blacklisted.
@@ -54,7 +72,7 @@ public interface Joiner {
      * @param address the address to check.
      * @return true if blacklisted, false otherwise.
      * @throws java.lang.NullPointerException if address is null.
-     * @see #blacklist(com.hazelcast.nio.Address)
+     * @see #blacklist(com.hazelcast.nio.Address, boolean)
      */
     boolean isBlacklisted(Address address);
 }

--- a/hazelcast/src/main/java/com/hazelcast/cluster/impl/AbstractJoiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/cluster/impl/AbstractJoiner.java
@@ -32,9 +32,8 @@ import com.hazelcast.util.Clock;
 
 import java.util.ArrayList;
 import java.util.Collection;
-import java.util.Collections;
-import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
@@ -50,8 +49,8 @@ public abstract class AbstractJoiner implements Joiner {
     private final ExceptionHandler whileWaitMergeExceptionHandler;
     private final AtomicLong joinStartTime = new AtomicLong(Clock.currentTimeMillis());
     private final AtomicInteger tryCount = new AtomicInteger(0);
-    protected final Set<Address> blacklistedAddresses
-            = Collections.newSetFromMap(new ConcurrentHashMap<Address, Boolean>());
+    // map blacklisted endpoints. Boolean value represents if blacklist is temporary or permanent
+    protected final ConcurrentMap<Address, Boolean> blacklistedAddresses = new ConcurrentHashMap<Address, Boolean>();
     protected final Config config;
     protected final Node node;
     protected final ILogger logger;
@@ -67,14 +66,23 @@ public abstract class AbstractJoiner implements Joiner {
     }
 
     @Override
-    public void blacklist(Address callerAddress) {
-        logger.info(callerAddress + " is added to the blacklist.");
-        blacklistedAddresses.add(callerAddress);
+    public void blacklist(Address address, boolean permanent) {
+        logger.info(address + " is added to the blacklist.");
+        blacklistedAddresses.putIfAbsent(address, permanent);
+    }
+
+    @Override
+    public boolean unblacklist(Address address) {
+        if (blacklistedAddresses.remove(address, Boolean.FALSE)) {
+            logger.info(address + " is removed from the blacklist.");
+            return true;
+        }
+        return false;
     }
 
     @Override
     public boolean isBlacklisted(Address address) {
-        return blacklistedAddresses.contains(address);
+        return blacklistedAddresses.containsKey(address);
     }
 
     public abstract void doJoin();

--- a/hazelcast/src/main/java/com/hazelcast/cluster/impl/TcpIpJoiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/cluster/impl/TcpIpJoiner.java
@@ -46,7 +46,7 @@ import static com.hazelcast.util.AddressUtil.AddressHolder;
 public class TcpIpJoiner extends AbstractJoiner {
 
     private static final int MAX_PORT_TRIES = 3;
-    public static final long JOIN_RETRY_WAIT_TIME = 1000L;
+    private static final long JOIN_RETRY_WAIT_TIME = 1000L;
 
     private volatile boolean claimingMaster = false;
 
@@ -119,8 +119,6 @@ public class TcpIpJoiner extends AbstractJoiner {
             Collection<Address> possibleAddresses = getPossibleAddresses();
 
             boolean foundConnection = tryInitialConnection(possibleAddresses);
-            removeBlacklistedAddresses(possibleAddresses);
-
             if (!foundConnection) {
                 logger.finest("This node will assume master role since no possible member where connected to.");
                 node.setAsMaster();
@@ -137,8 +135,7 @@ public class TcpIpJoiner extends AbstractJoiner {
                     return;
                 }
 
-                removeBlacklistedAddresses(possibleAddresses);
-                if (possibleAddresses.size() == 0) {
+                if (isAllBlacklisted(possibleAddresses)) {
                     logger.finest(
                             "This node will assume master role since none of the possible members accepted join request.");
                     node.setAsMaster();
@@ -150,8 +147,10 @@ public class TcpIpJoiner extends AbstractJoiner {
                     boolean consensus = claimMastership(possibleAddresses);
                     if (consensus) {
                         if (logger.isFinestEnabled()) {
+                            Set<Address> votingEndpoints = new HashSet<Address>(possibleAddresses);
+                            votingEndpoints.removeAll(blacklistedAddresses.keySet());
                             logger.finest("Setting myself as master after consensus! " +
-                                    "Voting endpoints: " + possibleAddresses);
+                                    "Voting endpoints: " + votingEndpoints);
                         }
                         node.setAsMaster();
                         claimingMaster = false;
@@ -172,6 +171,9 @@ public class TcpIpJoiner extends AbstractJoiner {
         claimingMaster = true;
         Collection<Future<Boolean>> responses = new LinkedList<Future<Boolean>>();
         for (Address address : possibleAddresses) {
+            if (isBlacklisted(address)) {
+                continue;
+            }
             if (node.getConnectionManager().getConnection(address) != null) {
                 Future future = node.nodeEngine.getOperationService()
                         .createInvocationBuilder(ClusterServiceImpl.SERVICE_NAME,
@@ -223,9 +225,8 @@ public class TcpIpJoiner extends AbstractJoiner {
         long start = Clock.currentTimeMillis();
 
         while (!node.joined() && Clock.currentTimeMillis() - start < connectionTimeoutMillis) {
-
             Address masterAddress = node.getMasterAddress();
-            if (possibleAddresses.isEmpty() && masterAddress == null) {
+            if (isAllBlacklisted(possibleAddresses) && masterAddress == null) {
                 return;
             }
 
@@ -235,7 +236,6 @@ public class TcpIpJoiner extends AbstractJoiner {
                 }
                 node.clusterService.sendJoinRequest(masterAddress, true);
             } else {
-                removeBlacklistedAddresses(possibleAddresses);
                 sendMasterQuestion(possibleAddresses);
             }
 
@@ -245,20 +245,11 @@ public class TcpIpJoiner extends AbstractJoiner {
         }
     }
 
-    private void removeBlacklistedAddresses(Collection<Address> possibleAddresses) {
-        if (!blacklistedAddresses.isEmpty()) {
-            if (logger.isFinestEnabled()) {
-                logger.finest("Removing failedConnections: " + blacklistedAddresses);
-            }
-            possibleAddresses.removeAll(blacklistedAddresses);
-        }
-    }
-
     private boolean tryInitialConnection(Collection<Address> possibleAddresses) throws InterruptedException {
         long connectionTimeoutMillis = getConnTimeoutSeconds() * 1000L;
         long start = Clock.currentTimeMillis();
         while (Clock.currentTimeMillis() - start < connectionTimeoutMillis) {
-            if (possibleAddresses.isEmpty()) {
+            if (isAllBlacklisted(possibleAddresses)) {
                 return false;
             }
             if (logger.isFinestEnabled()) {
@@ -268,9 +259,12 @@ public class TcpIpJoiner extends AbstractJoiner {
                 return true;
             }
             Thread.sleep(JOIN_RETRY_WAIT_TIME);
-            removeBlacklistedAddresses(possibleAddresses);
         }
         return false;
+    }
+
+    private boolean isAllBlacklisted(Collection<Address> possibleAddresses) {
+        return blacklistedAddresses.keySet().containsAll(possibleAddresses);
     }
 
     private void lookForMaster(Collection<Address> possibleAddresses) throws InterruptedException {
@@ -279,8 +273,7 @@ public class TcpIpJoiner extends AbstractJoiner {
             sendMasterQuestion(possibleAddresses);
             //noinspection BusyWait
             Thread.sleep(JOIN_RETRY_WAIT_TIME);
-            possibleAddresses.removeAll(blacklistedAddresses);
-            if (possibleAddresses.size() == 0) {
+            if (isAllBlacklisted(possibleAddresses)) {
                 break;
             }
         }
@@ -289,7 +282,7 @@ public class TcpIpJoiner extends AbstractJoiner {
             return;
         }
 
-        if (possibleAddresses.size() == 0 && node.getMasterAddress() == null) {
+        if (isAllBlacklisted(possibleAddresses) && node.getMasterAddress() == null) {
             if (logger.isFinestEnabled()) {
                 logger.finest("Setting myself as master! No possible addresses remaining to connect...");
             }
@@ -328,8 +321,14 @@ public class TcpIpJoiner extends AbstractJoiner {
     }
 
     private boolean sendMasterQuestion(Collection<Address> possibleAddresses) {
+        if (logger.isFinestEnabled()) {
+            logger.finest("NOT sending master question to blacklisted endpoints: " + blacklistedAddresses);
+        }
         boolean sent = false;
         for (Address address : possibleAddresses) {
+            if (isBlacklisted(address)) {
+                continue;
+            }
             if (logger.isFinestEnabled()) {
                 logger.finest("Sending master question to " + address);
             }

--- a/hazelcast/src/main/java/com/hazelcast/cluster/impl/operations/GroupMismatchOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/cluster/impl/operations/GroupMismatchOperation.java
@@ -24,7 +24,7 @@ public class GroupMismatchOperation extends AbstractClusterOperation
                 + " Cause: the target cluster has a different group-name");
 
         Node node = nodeEngine.getNode();
-        node.getJoiner().blacklist(getCallerAddress());
+        node.getJoiner().blacklist(getCallerAddress(), true);
     }
 }
 

--- a/hazelcast/src/main/java/com/hazelcast/nio/IOService.java
+++ b/hazelcast/src/main/java/com/hazelcast/nio/IOService.java
@@ -68,6 +68,8 @@ public interface IOService {
 
     ThreadGroup getThreadGroup();
 
+    void onSuccessfulConnection(Address address);
+
     void onFailedConnection(Address address);
 
     void shouldConnectTo(Address address);

--- a/hazelcast/src/main/java/com/hazelcast/nio/NodeIOService.java
+++ b/hazelcast/src/main/java/com/hazelcast/nio/NodeIOService.java
@@ -148,9 +148,16 @@ public class NodeIOService implements IOService {
     }
 
     @Override
+    public void onSuccessfulConnection(Address address) {
+        if (!node.joined()) {
+            node.getJoiner().unblacklist(address);
+        }
+    }
+
+    @Override
     public void onFailedConnection(Address address) {
         if (!node.joined()) {
-            node.getJoiner().blacklist(address);
+            node.getJoiner().blacklist(address, false);
         }
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpConnectionManager.java
+++ b/hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpConnectionManager.java
@@ -248,6 +248,7 @@ public class TcpIpConnectionManager implements ConnectionManager {
             return false;
         }
         connection.setEndPoint(remoteEndPoint);
+        ioService.onSuccessfulConnection(remoteEndPoint);
         if (reply) {
             sendBindRequest(connection, remoteEndPoint, false);
         }
@@ -314,6 +315,7 @@ public class TcpIpConnectionManager implements ConnectionManager {
 
     void sendBindRequest(TcpIpConnection connection, Address remoteEndPoint, boolean replyBack) {
         connection.setEndPoint(remoteEndPoint);
+        ioService.onSuccessfulConnection(remoteEndPoint);
         //make sure bind packet is the first packet sent to the end point.
         if (logger.isFinestEnabled()) {
             log(Level.FINEST, "Sending bind packet to " + remoteEndPoint);

--- a/hazelcast/src/test/java/com/hazelcast/test/TestNodeRegistry.java
+++ b/hazelcast/src/test/java/com/hazelcast/test/TestNodeRegistry.java
@@ -210,7 +210,12 @@ final class TestNodeRegistry {
         }
 
         @Override
-        public void blacklist(Address callerAddress) {
+        public void blacklist(Address address, boolean permanent) {
+        }
+
+        @Override
+        public boolean unblacklist(Address address) {
+            return false;
         }
 
         @Override

